src/backend/cpu/kernel/pad_array_borders.hpp

```diff
@@ -0,0 +1,119 @@
+/*******************************************************
+ * Copyright (c) 2018, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#pragma once
+#include <Param.hpp>
+
+namespace cpu
+{
+namespace kernel
+{
+static dim_t
+calcIdx4Border(const dim_t i, const dim_t lb,
```

Comment:
This function should really be optimized. This is being used a lot in your function and it is using some expensive operations (lots of branching and modulus). You should think more carefully about how to implement this.